### PR TITLE
std.typecons: make RefCounted @safe

### PR DESCRIPTION
**Pull request by wilzbach**
follow-up to #4685

### CODE REVIEW

File: std/typecons.d [hunk 1]
```diff
@@ -4892,7 +4892,7 @@ to deallocate the corresponding resource.
             GC.removeRange(&_refCounted._store._payload);
         }
         import core.stdc.stdlib : free;
-        free(_refCounted._store);
+        () @trusted { free(_refCounted._store); }();
```
Comment (wilzbach) [inline std/typecons.d]: I am not sure about this one. Is it safe to free memory from a pointer that we created and keep the reference too?

File: std/typecons.d [hunk 2]
```diff
@@ -7109,7 +7112,7 @@ private template replaceTypeInFunctionType(From, To, fun)
     struct S2 { void bar() { x = 2; } }
 
     alias Pass = Test!(
-        int, float, typeof(&func), float delegate(float),
+        int, float, typeof(&func), float delegate(float) @safe,
```
Comment (wilzbach) [inline std/typecons.d]: idk why this change was required, seemed like bug?

File: std/typecons.d [hunk 3]
```diff
@@ -4755,7 +4755,7 @@ if (!is(T == class) && !(is(T == interface)))
             import core.stdc.stdlib : malloc;
             import std.conv : emplace;
 
-            _store = cast(Impl*)malloc(Impl.sizeof);
+            _store = () @trusted { return cast(Impl*)malloc(Impl.sizeof); }();
```
Comment (wilzbach) [inline std/typecons.d]: afaict is allocation of new memory safe? Do we need more assumptions here?

File: std/typecons.d [hunk 4]
```diff
@@ -5010,7 +5010,10 @@ assert(refCountedStore.isInitialized)).
         assert(rc1._refCounted._store._count == 2);
     }
     assert(p._refCounted._store == null);
+}
```
Comment (schveiguy) [inline std/typecons.d]: This thing isn't technically `@safe`, because p is pointing at memory that is out of scope. You should leave this one as `@system`. What you can do is change the scope to be a `@safe` lambda. inside the `@system` unit test. Everything in here is safe except for this final check.

Comment (schveiguy) [inline std/typecons.d]: Er... not a lambda, you can probably make it an inner function.

**Comment by schveiguy** [comment]
What we also need is a check to make sure `RefCounted` becomes `@system` when a `@system` ctor or dtor is on the target type.

Thanks, this looks pretty good!

File: std/typecons.d (hunk 3 above)
Comment (schveiguy) [inline std/typecons.d]: manual memory management is safe as long as you properly pair the allocation with the deallocation. This is exactly what RefCounted does.

**Comment by schveiguy** [comment]
Since we never expose a direct pointer to the ref counted payload, I think `RefCounted` can be perfectly safe.

Note, there is an existing issue for this: https://issues.dlang.org/show_bug.cgi?id=13983

**Comment by jmdavis** [comment]
> Unless I'm missing something, we cannot currently make RefCounted @safe in that way since refCountedPayload, i.e. the AliasThis, returns a reference (which is the entire point of RefCounted). You'd need to make that @system to be able to mark the destructor @safe, which of course isn't any better.

Yes, and that's a good part of why Walter is currently working on adding a ref-counting solution to the language - because while you can make ref-counting solution with a library, you can't make an `@safe` ref-counting solution with a library.

**Comment by schveiguy** [comment]
@klickverbot as long as you are in safe code, using the reference has to be while the ref-counted struct exists. You can't take the address of a ref in `@safe` code, so how is that unsafe?

**Comment by schveiguy** [comment]
Nevermind, I just realized that you can *return* a ref, so that would be bad. Yeah, we have to wait on this.

**Comment by nordlow** [comment]
Why is `RefCounted` not `pure`?

Because of `malloc`, `free`, etc?
